kframework/article.md

```diff
@@ -0,0 +1,169 @@
+
+To find out if the functions in question are getting inlined, we can insert the following line to our module:
+
+```hs
+{-# OPTIONS -ddump-simpl -ddump-to-file #-}
+```
+
+1. `-ddump-to-file` causes the output from each of flags starting with `-ddump`, to be dumped to a file or files.
+2. `-ddump-simpl` dumps simplifier output.
+
+After building our module, GHC will produce a `.dump-simpl` file containing its internal representation of the program after most optimizations have had the chance to take place.
+
+Suppose we built the module where the `do`-notation snippet is defined, and we worked in an abstract monad `m`. When we search for `>>=` in our generated `.dump-simpl` file, we will stumble upon this snippet:
+
+```
+>>=
+  @ m_aM5
+  $dMonad_aMr
+  @ a_aM6
+  @ c_aM8
+  eta1_B2
+  (\ (a1_anu :: a_aM6) ->
+     >>=
+       @ m_aM5
+       $dMonad_aMr
+       @ b_aM7
+       @ c_aM8
+       eta2_B1
+       (\ (b1_anv :: b_aM7) -> eta_B3 a1_anu b1_anv))
+```
+
+The call to `>>=` is **not** inlined, and the reason is because it is instantiated at an abstract monad `m_aM5`, and GHC does not now which instance to use. Instead, it has to pass the implementation in a dictionary `$dMonad_aMr`.
+
+## Findings
+
+Here is a list of relevant terminology, datatypes, and typeclasses:
+* **Satisfiability Modulo Theories** (**SMT** for short) is the problem of determining whether a first-order formula is satisfiable, i.e. if it is true under a certain assignment of values to its variables, and acts as the generalization of the Boolean satisfiability problem (SAT).
+* An **SMT solver** is a tool that aims to solve the SMT problem in given theories and a subset of inputs.
+* `SMT a` is a datatype that queries an external SMT solver.
+* `NoSMT a` is a datatype for the case when no SMT solver is provided.
+* `MonadSMT m` is a typeclass that allows accessing `SMT` through monad transformers. It has two base instances: `MonadSMT SMT` and `MonadSMT NoSMT`.
+* `SimplifierT smt a` is a datatype representing a simplification action, the `smt` type variable can be subsittuted with either `SMT` or `NoSMT`.
+* `MonadSimplify m` is a typeclass containing functions relevant to the simplification process. Allows working with arbitrarily nested monad transformers. Its base instance is `MonadSimplify (SimplifierT m)`.
+
+### Specialization
+
+We attempted to solve the [aforementioned problem of inlining](#inlining) by changing the abstract monad `MonadSimplify simplifier` to `SimplifierT SMT`. This could not be done easily because of the application architecture: the base monad, `SMT` or `NoSMT`, is determined dynamically.
+
+```
+execute options metadataTools lemmas worker =
+  case solver of
+    Z3 -> withZ3
+    None -> withoutSMT
+```
+
+For the purposes of this investigation we removed the support of `NoSMT`.
```

Comment:
We didn't completely remove it though. `withoutSMT` still [exists](https://github.com/runtimeverification/haskell-backend/blob/a1959233e210f960d49a91ebe7fcec76da14ebe3/kore/app/share/GlobalMain.hs#L558). We have changed the implementation so that the information where SMT solver is provided is stored inside runtime environment.

Reply:
This is the phrase that I copied from Vladislav's performance report when the discovery phase was conducted back in May 2022. I think you're referring to the ultimate fate of the SMT datatype when we had merged the two by wrapping the environment in the Maybe context.

Perhaps, I can restructure this sentence a little bit, to make sure the reader is able to unambiguously understand that this maneuver is performed specifically in the discovery phase. What do you think?

Reply:
Yes, sounds good to me.

Reply:
@breakerzirconia this sentence still needs to be updated.